data_to_long

DominiqueMakowski opened the pull request:
For https://github.com/easystats/parameters/issues/482

Will also be useful for bayestestR::reshape_iterations() 

Potential improvements include:

- [ ] Support more of the pivot_longer arguments (e.g., prefixes etc)
- [ ] More flexible col selection (currently only supports vector of names or indices), but it'd be nice to support regexes, NS names etc.

DominiqueMakowski commented:
``` r
library(insight)

data <- data.frame(replicate(5, rnorm(10)))

# Default behaviour (equivalent to tidyr::pivot_longer(data, cols = 1:5))
data_to_long(data)
#>    Name        Value
#> 1    X1 -0.479473986
#> 2    X2 -0.152437350
#> 3    X3  1.128432563
#> 4    X4 -0.102928292
#> 5    X5 -0.237663283
#> 6    X1 -0.446248983
#> 7    X2  1.346009889
#> 8    X3 -0.392110719
#> 9    X4 -0.683979861
#> 10   X5 -1.169995448
#> 11   X1 -0.674086138
#> 12   X2 -2.998579713
#> 13   X3 -1.357403961
#> 14   X4 -1.188147055
#> 15   X5  0.531296987
#> 16   X1  2.085650837
#> 17   X2 -1.221659956
#> 18   X3 -1.987809126
#> 19   X4 -0.472785483
#> 20   X5 -1.183716098
#> 21   X1  0.160160273
#> 22   X2 -1.249709842
#> 23   X3  1.449698949
#> 24   X4  0.452405709
#> 25   X5  0.109714165
#> 26   X1 -0.242604977
#> 27   X2 -0.372166722
#> 28   X3 -2.622868216
#> 29   X4  0.076762922
#> 30   X5 -1.258398395
#> 31   X1 -1.805719582
#> 32   X2 -0.573946488
#> 33   X3 -0.150443077
#> 34   X4 -0.006815657
#> 35   X5 -0.145239995
#> 36   X1  0.150729438
#> 37   X2  2.095734621
#> 38   X3  0.154003233
#> 39   X4 -1.133291221
#> 40   X5 -2.512204451
#> 41   X1  1.050840116
#> 42   X2 -1.578521404
#> 43   X3 -0.358654378
#> 44   X4  1.161663135
#> 45   X5  0.689856266
#> 46   X1  0.332503557
#> 47   X2 -0.738281019
#> 48   X3 -0.557748169
#> 49   X4 -0.724874458
#> 50   X5 -0.952063226

# Customizing the names
data_to_long(data,
             cols = c(1, 2),
             colnames_to = "Column",
             values_to = "Numbers",
             rows_to = "Row")
#>            X3           X4         X5 Row Column    Numbers
#> 1   1.1284326 -0.102928292 -0.2376633   1     X1 -0.4794740
#> 2   1.1284326 -0.102928292 -0.2376633   1     X2 -0.1524374
#> 3  -0.3921107 -0.683979861 -1.1699954   2     X1 -0.4462490
#> 4  -0.3921107 -0.683979861 -1.1699954   2     X2  1.3460099
#> 5  -1.3574040 -1.188147055  0.5312970   3     X1 -0.6740861
#> 6  -1.3574040 -1.188147055  0.5312970   3     X2 -2.9985797
#> 7  -1.9878091 -0.472785483 -1.1837161   4     X1  2.0856508
#> 8  -1.9878091 -0.472785483 -1.1837161   4     X2 -1.2216600
#> 9   1.4496989  0.452405709  0.1097142   5     X1  0.1601603
#> 10  1.4496989  0.452405709  0.1097142   5     X2 -1.2497098
#> 11 -2.6228682  0.076762922 -1.2583984   6     X1 -0.2426050
#> 12 -2.6228682  0.076762922 -1.2583984   6     X2 -0.3721667
#> 13 -0.1504431 -0.006815657 -0.1452400   7     X1 -1.8057196
#> 14 -0.1504431 -0.006815657 -0.1452400   7     X2 -0.5739465
#> 15  0.1540032 -1.133291221 -2.5122045   8     X1  0.1507294
#> 16  0.1540032 -1.133291221 -2.5122045   8     X2  2.0957346
#> 17 -0.3586544  1.161663135  0.6898563   9     X1  1.0508401
#> 18 -0.3586544  1.161663135  0.6898563   9     X2 -1.5785214
#> 19 -0.5577482 -0.724874458 -0.9520632  10     X1  0.3325036
#> 20 -0.5577482 -0.724874458 -0.9520632  10     X2 -0.7382810
```

<sup>Created on 2021-05-09 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

IndrajeetPatil commented:
Nice! These functions will also be helpful for other developers, not just for `easystats` functions, so glad to see them!

One of the common inconsistencies across packages that deal with statistical testing is whether functions use long or wide format data for _paired or repeated measures_ designs. I sometimes need to convert repeated measures data in long format to wide format, which is really pain in the ass if there are missing values which need to be excluded. This is because such exclusion needs to take into account the within-subjects nature of the data into account (I do have a helper function that lets me do this now, but it took a while to get it right: https://github.com/IndrajeetPatil/ipmisc/blob/master/R/long_to_wide_converter.R).

IndrajeetPatil commented:
> Is it possible to add that feature to our new functions?

I can play around with them later and see if we can improve on them, but don't hold off merging the PR on account of that.

DominiqueMakowski commented:
also needed to add `data_match()` to quickly find subsets of a dataframe that match a particular specification:

``` r
mtcars[insight::data_match(mtcars, data.frame(vs = 0, am = 1)), ]
#>                 mpg cyl  disp  hp drat    wt  qsec vs am gear carb
#> Mazda RX4      21.0   6 160.0 110 3.90 2.620 16.46  0  1    4    4
#> Mazda RX4 Wag  21.0   6 160.0 110 3.90 2.875 17.02  0  1    4    4
#> Porsche 914-2  26.0   4 120.3  91 4.43 2.140 16.70  0  1    5    2
#> Ford Pantera L 15.8   8 351.0 264 4.22 3.170 14.50  0  1    5    4
#> Ferrari Dino   19.7   6 145.0 175 3.62 2.770 15.50  0  1    5    6
#> Maserati Bora  15.0   8 301.0 335 3.54 3.570 14.60  0  1    5    8
```

<sup>Created on 2021-05-09 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

DominiqueMakowski commented:
I also moved `data_restoretype` that "fixes" if need be a dataframe according to the column types of another, typically by restoring factor levels etc. Can be useful when we extract data from models or generate new data, as often the types can be changed.

``` r
data <- data.frame(
  Sepal.Length = c("1", "3"),
  Species = c("setosa", "versicolor"),
  New = c("1", "3")
)

fixed <- insight::data_restoretype(data, reference = iris)
summary(fixed)
#>   Sepal.Length       Species      New           
#>  Min.   :1.0   setosa    :1   Length:2          
#>  1st Qu.:1.5   versicolor:1   Class :character  
#>  Median :2.0   virginica :0   Mode  :character  
#>  Mean   :2.0                                    
#>  3rd Qu.:2.5                                    
#>  Max.   :3.0
levels(fixed$Species)
#> [1] "setosa"     "versicolor" "virginica"
```

<sup>Created on 2021-05-09 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

bwiernik commented:
Could we call it `reshape_longer()`? Two thoughts why:

1) I find the multiple underscores awkward to type, and "reshape" follows the base R name and convention of action functions being verbs. 

2) It avoids a dichotomous notion of "long" and "wide" data. Could be very long or very wide, etc.

bwiernik commented:
> also needed to add data_match() to quickly find subsets of a dataframe that match a particular specification:

Isn't this just `subset()`?

bwiernik commented:
My general comment on these is that I think we should ideally match the API of the tidyr::pivot functions exactly (or at least be a superset of it if we have additional features).

DominiqueMakowski commented:
> reshape_longer()

That would also fit with our API since we already have `reshape_iterations()`. Should we have these as aliases?

> Isn't this just subset()?

If I'm not mistaken data_match() takes partial dataframe as match and returns indices?

> we should ideally match the API of the tidyr::pivot functions exactly

So far we have the same args names. I just allowed for `names_to/from` to be additionally specified explicitly as `colnames_to/from` (as it's a common synonym in R - `colnames() / names()`)

DominiqueMakowski commented:
I leave it to @IndrajeetPatil ^^
aaaaand also because we need it in `parameters` 😁

IndrajeetPatil commented:
I can work on this once `easystats` meta-package is on CRAN.

Shuffling functions around is going to be a bit of a headache, so would like to do this once the API as stable as possible and the code as thoroughly tested as possible.

DominiqueMakowski commented:
> Shuffling functions around is going to be a bit of a headache, so would like to do this once the API as stable as possible and the code as thoroughly tested as possible

yeah makes sense!
